core/arch/arm/mm/core_mmu.c

```diff
@@ -988,6 +988,78 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
 	}
 }
 
+TEE_Result core_mmu_map_pages(vaddr_t start, paddr_t *pages, size_t num_pages,
+			      enum teecore_memtypes memtype)
+{
+	TEE_Result ret;
+	struct core_mmu_table_info tbl_info;
+	unsigned int idx;
+	uint32_t old_attr;
+	vaddr_t vaddr = start;
+	size_t i;
+
+	if (vaddr & SMALL_PAGE_MASK)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	for (i = 0; i < num_pages; i++) {
+		if (pages[i] & SMALL_PAGE_MASK) {
+			ret = TEE_ERROR_BAD_PARAMETERS;
+			goto out;
+		}
+
+		while (true) {
+			if (!core_mmu_find_table(vaddr, UINT_MAX, &tbl_info)) {
+				ret = TEE_ERROR_OUT_OF_MEMORY;
+				goto out;
+			}
+
+			idx = core_mmu_va2idx(&tbl_info, vaddr);
+			if (tbl_info.shift != SMALL_PAGE_SHIFT) {
+				if (!core_mmu_divide_block(&tbl_info, idx)) {
+					EMSG("Could not divide block into smaller tables");
+					ret = TEE_ERROR_OUT_OF_MEMORY;
+					goto out;
+				}
+			} else {
+				break;
+			}
+		}
+
+		core_mmu_get_entry(&tbl_info, idx, 0, &old_attr);
+		if (old_attr)
+			panic("Page is already mapped\n");
+
+		core_mmu_set_entry(&tbl_info, idx, pages[i],
+				   core_mmu_type_to_attr(memtype));
+		vaddr += SMALL_PAGE_SIZE;
+	}
+
+	core_tlb_maintenance(TLBINV_UNIFIEDTLB, 0);
```

Comment:
This shouldn't be needed as we've ensured that nothing was mapped in the va range before.